packages/js/components/src/slot-context/slot-context.tsx

```diff
@@ -0,0 +1,67 @@
+/**
+ * External dependencies
+ */
+import { createElement, createContext, useContext } from '@wordpress/element';
+
+type FillConfigType = {
+	visible: boolean;
+};
+
+type FillType = Record< string, FillConfigType >;
+
+type FillCollection = readonly ( readonly JSX.Element[] )[];
+
+export type SlotContextHelpersType = {
+	hideFill: ( id: string ) => void;
+	showFill: ( id: string ) => void;
+	getFills: () => FillType;
+};
+
+export type SlotContextType = {
+	fills: FillType;
+	getFillHelpers: () => SlotContextHelpersType;
+	registerFill: ( id: string ) => void;
+	filterRegisteredFills: ( fillsArrays: FillCollection ) => FillCollection;
+};
+
+export const SlotContextPrototype: SlotContextType = ( () => {
```

Comment:
My thought was to keep this as a generic component that could be used for other areas of the site. This would only store the fills information for one particular area, which would be governed by the placement of the provider. In this instance the provider is just encapsulating the `<ProductForm />` so it would only include product editor fills. 

I'm glad to add an ID of some sort, but I'm not sure we'd actually use it for anything logic-wise at the moment...?

Reply:
Gotcha, yeah an `ID` might not work that well then and keeping this generic would be better.
Creating separate providers for each area would fix the issue of triggering a `re-render` in unnecessary places.
Although I do wonder if it becomes confusing to use the generic `hideFill` & `showFill` functions when not aware of which provider it's actually being called for.
Mostly calling this out as we could have multiple of these on a single page, but no way of telling which provider we are in or is being called.
Maybe some form of `id` would be beneficial here, or a wrapper Context class that is more specific.

Reply:
Right, I think it depends a little on how broad we want to make these contexts. My initial thought was that the context would encompass the entire product form, so any fills within the product editor could manage other compatible fills. But, if we were to use this on future Settings page(s) for example, that would be a separate context. 

I _think_ that would just align with the "expected" behavior. The consumers that are using the fills wouldn't need to really know about the architecture of the context, but would expect that they're only privy to other fills on the product editor.

Reply:
We could pass an ID and just make that available as a passed prop or something that allows fills to verify which context they're using, although if we use something like the above it should feel fairly obvious that you're just seeings fills on the same page/screen.